File: src/com/github/stromland/piano/services/TriadChord.kt
package com.github.stromland.piano.services

import com.github.stromland.piano.models.PianoKey
import com.github.stromland.piano.utils.sliceToLast

object TriadChord {

    fun inversions(): List<String> {
        return listOf("root", "firstInversion", "secondInversion")
    }

    fun findChord(pianoKeys: List<PianoKey>, scaleKeys: List<Int>, startKey: Int): Chord {
        return triadChords().map { chord ->
            Chord(
                pianoKeys[startKey].note,
                chord.value.suffix,
                chord.value.keyIndexes.map { it + startKey }
            )
        }.first { it.keyIndexes.all { scaleKeys.contains(it) } }
    }

    fun getTriadInversions(chordIndexes: List<Int>): List<List<Int>> {
        fun shiftIndexes(x: Int): List<Int> {
            return (List(x) { 12 }.mapIndexed { i, v -> chordIndexes[i] + v } + chordIndexes.sliceToLast(x)).sorted()
        }
        return listOf(chordIndexes, shiftIndexes(1), shiftIndexes(2))
    }

    private fun triadChords(): Map<String, ChordSpec> {
        return mapOf(
            "major" to ChordSpec("", listOf(0, 4, 7)),
            "minor" to ChordSpec("m", listOf(0, 3, 7)),
            "diminished" to ChordSpec("dim", listOf(0, 3, 6)),
            "augmented" to ChordSpec("aug", listOf(0, 4, 8))
        )
    }

    data class Chord(val note: String, val suffix: String, val keyIndexes: List<Int>)

    data class ChordSpec(val suffix: String, val keyIndexes: List<Int>)
}

File: src/services/TriadChord.ts
import { PianoKey } from "models/PianoKey";
import { all } from "utils/functions";

export interface Chord extends ChordSpec {
  note: string;
}

export interface ChordSpec {
  suffix: string;
  keyIndexes: number[];
}

export enum Inversion {
  ROOT,
  FIRST,
  SECOND
}

interface TriadChordSpecs {
  major: ChordSpec;
  minor: ChordSpec;
  diminished: ChordSpec;
  augmented: ChordSpec;
  [key: string]: ChordSpec;
}

export class TriadChord {
  private static triadChordSpecs: TriadChordSpecs = {
    major: { suffix: "", keyIndexes: [0, 4, 7] },
    minor: { suffix: "m", keyIndexes: [0, 3, 7] },
    diminished: { suffix: "dim", keyIndexes: [0, 3, 6] },
    augmented: { suffix: "aug", keyIndexes: [0, 4, 8] }
  };

  public static findChord(
    pianoKeys: PianoKey[],
    scaleKeys: number[],
    startKey: number
  ): Chord | undefined {
    return Object.values<ChordSpec>(this.triadChordSpecs)
      .map(
        (spec): Chord => ({
          note: pianoKeys[startKey].note,
          suffix: spec.suffix,
          keyIndexes: spec.keyIndexes.map(it => it + startKey)
        })
      )
      .find(chord => all(chord.keyIndexes, key => scaleKeys.includes(key)));
  }

  public static getTriadInversion(
    chord: Chord,
    inversion: Inversion
  ): number[] {
    const shiftIndexes = (x: number) =>
      Array(x)
        .fill(12)
        .map((v, i) => chord.keyIndexes[i] + v)
        .concat(chord.keyIndexes.slice(x))
        .sort((x, y) => x - y);

    switch (inversion) {
      case Inversion.ROOT:
        return chord.keyIndexes;
      case Inversion.FIRST:
        return shiftIndexes(1);
      case Inversion.SECOND:
        return shiftIndexes(2);
    }
  }
}
